scrape/target.go

```diff
@@ -422,41 +399,54 @@ func populateLabels(lset labels.Labels, cfg *config.ScrapeConfig) (res, orig lab
 		lb.Set(model.InstanceLabel, addr)
 	}
 
-	res = lb.Labels()
+	res := lb.Labels()
 	for _, l := range res {
 		// Check label values are valid, drop the target if not.
 		if !model.LabelValue(l.Value).IsValid() {
-			return nil, nil, errors.Errorf("invalid label value for %q: %q", l.Name, l.Value)
+			return nil, errors.Errorf("invalid label value for %q: %q", l.Name, l.Value)
 		}
 	}
-	return res, preRelabelLabels, nil
+	return res, nil
 }
 
-// targetsFromGroup builds targets based on the given TargetGroup and config.
-func targetsFromGroup(tg *targetgroup.Group, cfg *config.ScrapeConfig) ([]*Target, error) {
-	targets := make([]*Target, 0, len(tg.Targets))
+// stackLabels stacks labels from different sources for a particular target.
+// Sequence of adding labels is very important.
+// Each set operation in labels.Builder OVERRIDES previous label value.
+// Sequence should be:
+// 1. Config scrape labels.
+// 2. Config query params labels.
+// 3. Common labels for a target group.
+// 4. Labels for a particular target.
+func stackTargetLabels(tlset model.LabelSet, tglset model.LabelSet, cfg *config.ScrapeConfig) labels.Labels {
+	lb := labels.NewBuilder(make([]labels.Label, 0, len(tlset)+len(tglset)))
+
+	// Scrape config labels. 
```

Comment:
There's no such thing as a scrape config label. We have fields in the scrape config that happen to become labels, but they have no special name.